torch/optim/optimizer.py

```diff
@@ -527,13 +619,13 @@ def add_param_group(self, param_group):
             if not isinstance(param, torch.Tensor):
                 raise TypeError("optimizer can only optimize Tensors, "
                                 "but one of the params is " + torch.typename(param))
-            if not self.defaults.get('differentiable', None) and not (param.is_leaf or param.retains_grad):
+            # TODO: https://github.com/pytorch/pytorch/issues/30788
```

Comment:
this TODO seems irrelevant

Reply:
I added this TODO to explain the `type: ignore[attr-defined]` below since `retains_grad ` is not exposed even though it is available and being used. I can remove it if you want.

Reply:
I added a clarifying comment about what this is for LMK if you still want me to remove it.

Reply:
Instead of the TODO linking to just the issue, could you write a comment instead on why the type ignore is necessary? (e.g., why is it not exposed/where is it used/when we can expect to resolve this). Feel free to still reference the issue but it is not clear why the issue is relevant from having no context.

Reply:
Okay so this is what I am going to put instead:

```python
# Tensor.retains_grad is not exposed in the Python API https://github.com/pytorch/pytorch/issues/30788
```

Reply:
I'm still failing to see the connection between retains_grad typing and the Jit/C++ issue. Could you elaborate on how they're related?

Reply:
I think I misunderstood the purpose of that ticket. My goal was to say that we need to at add `retains_grad` to the definition of `_TensorBase` in the generated  `torch/_C/__init__.pyi`, otherwise static type-checkers have no idea that this attribute exists on `Tensor`s (like we see here).

Reply:
I'm not sure when or why it exists or if that is the appropriate place to add it or how it would get added, though.

Reply:
I went ahead and removed it but I am curious if you know how we can expose it

Reply:
Oh I see, yea the other issue is a really old issue tracking compatibility with JIT...which is now not really maintained. Feel free to add a comment saying "we need to at add retains_grad to the definition of _TensorBase in the generated torch/_C/__init__.pyi, otherwise static type-checkers have no idea that this attribute exists on Tensors" explaining the type ignore

Reply:
https://github.com/pytorch/pytorch/issues/103485

Reply:
Beautiful